h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,195 @@
+     * @param responseLevelIndex
+     */
+    private static void validateArgs(TreeV3 args, SharedTreeModel.SharedTreeOutput output, final int responseLevelIndex) {
+        if (args.tree_number < 0) throw new IllegalArgumentException("Tree number must be greater than 0.");
+
+        if (args.tree_number > output._treeKeys.length - 1)
+            throw new IllegalArgumentException("There is no such tree.");
+
+        if (responseLevelIndex < 0)
+            throw new IllegalArgumentException("There is no such tree class. Given categorical level does not exist in response column: " + args.tree_class.trim());
+    }
+
+
+    /**
+     * Converts H2O-3's internal representation of a boosted tree in a form of {@link SharedTreeSubgraph} to a format
+     * expected by H2O clients.
+     *
+     * @param sharedTreeSubgraph An instance of {@link SharedTreeSubgraph} to convert
+     * @return An instance of {@link TreeProperties} with some attributes possibly empty if suitable. Never null.
+     */
+    static TreeProperties convertSharedTreeSubgraph(final SharedTreeSubgraph sharedTreeSubgraph) {
+        Objects.requireNonNull(sharedTreeSubgraph);
+
+        final TreeProperties treeprops = new TreeProperties();
+
+        treeprops._leftChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        treeprops._rightChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        treeprops._descriptions = new String[sharedTreeSubgraph.nodesArray.size()];
+        treeprops._thresholds = MemoryManager.malloc4f(sharedTreeSubgraph.nodesArray.size());
+        treeprops.features = new String[sharedTreeSubgraph.nodesArray.size()];
+        treeprops.nas = MemoryManager.mallocZ(sharedTreeSubgraph.nodesArray.size());
+
+        // Set root node's children, there is no guarantee the root node will be number 0
+        treeprops._rightChildren[0] = sharedTreeSubgraph.rootNode.getRightChild() != null ? sharedTreeSubgraph.rootNode.getRightChild().getNodeNumber() : -1;
+        treeprops._leftChildren[0] = sharedTreeSubgraph.rootNode.getLeftChild() != null ? sharedTreeSubgraph.rootNode.getLeftChild().getNodeNumber() : -1;
+
+        List<SharedTreeNode> nodesToTraverse = new ArrayList<>();
+        nodesToTraverse.add(sharedTreeSubgraph.rootNode);
+        append(treeprops._rightChildren, treeprops._leftChildren,
+                treeprops._descriptions, treeprops._thresholds, treeprops.features, treeprops.nas,
+                nodesToTraverse, -1, false);
+
+        return treeprops;
+    }
+
+    private static void append(final int[] rightChildren, final int[] leftChildren, final String[] nodesDescriptions,
+                               final float[] thresholds, final String[] splitColumns, final boolean[] naHandlings,
+                               List<SharedTreeNode> nodesToTraverse, int pointer, boolean visitedRoot) {
```

Comment:
you can simplify method signature here: why not 
`append(TreeProperties treeprop, List<SharedTreeNode> nodesToTraverse, int pointer, boolean visitedRoot)` ?
won't prevent you from using recursion

Reply:
also, in your current usage of this method, pointer and visitedRoot seems to be redundant:
basically `visitedRoot = pointer > 0`